sdk/go/pulumi/transformation.go

```diff
@@ -0,0 +1,33 @@
+package pulumi
+
+// ResourceTransformationArgs is the argument bag passed to a resource transformation.
+type ResourceTransformationArgs struct {
+	// The resource instance that is being transformed.
+	Resource Resource
+	// The type of the resource.
+	Type string
+	// The name of the resource.
+	Name string
+	// The original properties passed to the resource constructor.
+	Props Input
```

Comment:
Is this the appropriate type? Both python and node use a mapping of str-> input:

https://github.com/pulumi/pulumi/blob/9abcca345a6c024087b35152b0d6fca92baef191/sdk/nodejs/resource.ts#L557
https://github.com/pulumi/pulumi/blob/33b232adc4737144c9e8626343050059cec032a7/sdk/python/lib/pulumi/resource.py#L232

Reply:
In Go, the concrete type here will typically be an Args struct - so definitely this is intentionally not the same as TypeScript/Go.

I was actually wondering something similar though.  It might be nice to have a test case that actually reads the props and creates a new one.  I assume it would need to cast from this type down into the concrete type expected as input for the resource?   I didn't see any tests covering that - but it will be a reasonably common usage mode of this feature.

Reply:
@pgavlin was trying to help me modify a `pulumi.Input`, but we ran into some issues when trying to run this snippet:
```
props := pulumi.All(args.Props).ApplyT(func(arr []interface{}) *testResourceArgs {
					props := arr[0].(*testResourceArgs)
					props.Foo = "baz"
					return props
				})
```
He thinks it's a bug in `awaitInputs` and started a debugging branch [here](https://github.com/pulumi/pulumi/compare/pgavlin/debugAwait). Simply put, I don't think the suggested test is possible until we figure out what's wrong with `awaitInputs`, which is why I straight up replace `props` in my tests rather than modifying them. If there's any alternative ways to do this, please let me know!

Reply:
Interesting.  Is it possible to at least read a value out of an `Input` and then construct a new `Input` that includes part of the initial one? 

This is morally similar to the following from the TypeScript tests which doesn't mutate, but does "copy over" properties from the incoming properties.  (There is of course no equivalent to the `...` in Go, we would have to enumerate all the properties of the Args type).

```ts
                 return {
                    props: { optionalInput: "newDefault", ...props },
                    opts: pulumi.mergeOptions(opts, { additionalSecretOutputs: ["output"] }),
                };
```

Reply:
To be honest, I'm not super clear on how `pulumi.Input` works, so I don't really know how to extract values from it, outside of what Pat showed me with that snippet above. Maybe @pgavlin or @EvanBoyle can give us more insight?

Reply:
@pgavlin I would have hoped there was an easy way to cast type `Props` had here directly to an `Args` object.  Is that not the case?  Similar to how in .NET this is:

```cs
public ResourceArgs Args { get; }
```

Which is a base type for all Args objects such that casting to the known actual arg type is straightforward.

The `pulumi.All(args.Props).ApplyT(func(arr []interface{}) *testResourceArgs {` dance above does not look ideal here.

Reply:
So there's no way I can see for the type system to enforce that the `Input` to `RegisterResource` is not an `Output`, but IIRC we do have runtime checks to ensure that it is a pointer to a struct, and we could extend those to also ensure that it is not an `Output`. This would allow users to assume that the type is a top-level `Args` type, and we could write this:

```
props := args.Props.(*testResourceInputs)
props.Foo = pulumi.String("baz")
```

@tasiah can you try changing your test to that approach?